components/nimbus/android/src/main/java/org/mozilla/experiments/nimbus/FeatureVariables.kt

```diff
@@ -15,53 +17,114 @@ import org.json.JSONObject
  * the value is of the correct type, then it is returned. If neither of these are true, then `null`
  * is returned.
  *
- * Supported types:
+ * ## Supported types:
  *
- * Basic:
+ * ### Primitive types:
  *
  *  - `String`
  *  - `Int`
  *  - `Boolean`
  *
- * Resource types. These use `getString` to look up an experiment value, then the app's `Context` to
- * find a resource.
+ * ### Types coerced from `String` values:
  *
- *  - `StringResource`
- *  - `DrawableResource`
+ *  - `Enum<T>`
+ *  - Resources
+ *      These use `getString` to look up an experiment value, then the app's `Context` to
+ *      find a resource. These are `Text` and `Drawable`.
  *
  * ```
  * val config = nimbus.getVariables("submitButton")
  *
  * submitButton.text = config.getText("submitButton.text") ?: R.string.submit_button_label
- * submitButton.color = config.getColor("submitButton.color") ?: R.color.button_default
- *
+ * submitButton.drawable = config.getDrawable("submitButton.color") ?: context.getDrawable(R.drawable.button_default)
  * ```
  *
+ * ### Nested `Variables`
+ *
+ * As `JSONObject`s can contain other `JSONObject`s, then so `Variables` can contain other `Variables`.
+ *
+ * Convenience methods are provided to map these inner Variables in to richer types.
+ *
+ * ### Structural types
+ *
+ * For all types, corresponding `List` and `Map` methods are available.
+ *
+ * In the case of all Maps, they are returned as `Map<String, T>`.
+ *
+ * ### Enums
+ *
+ * String coercion to Enums are supported, but provided as extension methods to
+ *
+ *  - `Variables.getEnumList<E>`,
+ *  - `Map<K, String>.mapKeysAsEnum`
+ *  - `Map<String, V>.mapValuesAsEnum`
+ *
+ * Special
+ *
  * Each of the keys needed by feature should be documented in the app's experiment manifest, which
  * will provide enough information the Experimenter to design an experiment.
  */
 interface Variables {
+    companion object {
+        val empty = NullVariables.instance
+    }
     /**
      * Finds a string typed value for this key. If none exists, `null` is returned.
      *
      * N.B. the `key` and type `String` should be listed in the experiment manifest.
      */
     fun getString(key: String): String? = null
 
+    /**
+     * Find an array for this key, and returns all the strings in that array. If none exists, `null`
+     * is returned.
+     */
+    fun getStringList(key: String): List<String>? = null
```

Comment:
Hmm is it possible to have this (and the other `get*List`) as one `getList<T: SomeInterface>(key: String): List<T>? = null` generic function?

I took a look at the implementations, and each `get*List` seems different enough to warrant the different functions, so feel free to completely ignore this 😅

Reply:
Agreed. I think I might be able to get the `getBoolList`, `getIntList` and `getStringList` to accept a type, however, I didn't because: 

 * there are enough restrictions from the kotlin compiler (see `reified T`) to make it non-straightforward.
 * many of the fundamental types are being coerced from Strings (e.g. Enums, Text, Drawable).
 * `getStringList` and `getTextList` have the same return type.

For this reason, I left it as `get*`, `get*List` and `get*Map`.